trie/database.go

```diff
@@ -712,7 +718,9 @@ func (db *Database) commit(hash common.Hash, batch ethdb.Batch, uncacher *cleane
 			return err
 		}
 		db.lock.Lock()
-		batch.Replay(uncacher)
+		if err := batch.Replay(uncacher); err != nil {
+			return err
```

Comment:
This is not correct because the lock will be held after returning. It could be fixed like:

```go
db.lock.Lock()
err := batch.Replay(uncacher)
batch.Reset()
db.lock.Unlock()
if err != nil {
    return err 
}
```

There is another instance of this mistake in the diff.

Reply:
Shouldn't it be something like this?

```go
db.lock.Lock()
err := batch.Replay(uncacher)
if err != nil {
    db.lock.Unlock()
    return err 
}
batch.Reset()
```

Reply:
It doesn't really matter if we reset the batch first, and doing so keeps the lock handling cleaner.